internal/controller/apiextensions/claim/syncer_ssa.go

```diff
@@ -74,32 +72,68 @@ func NewPatchingManagedFieldsUpgrader(w client.Writer) *PatchingManagedFieldsUpg
 // apply.
-func (u *PatchingManagedFieldsUpgrader) Upgrade(ctx context.Context, obj client.Object, ssaManager string, csaManagers ...string) error {
-	// UpgradeManagedFieldsPatch removes or replaces the specified CSA managers.
-	// Unfortunately most Crossplane controllers use CSA manager "crossplane".
-	// So we could for example fight with the XR controller:
-	//
-	// 1. We remove CSA manager "crossplane", triggering XR controller watch
-	// 2. XR controller uses CSA manager "crossplane", triggering our watch
-	// 3. Back to step 1 :)
-	//
-	// In practice we only need to upgrade once, to ensure we don't share fields
-	// that only this controller has ever applied with "crossplane". We assume
-	// that if our SSA manager already exists, we've done the upgrade.
-	for _, e := range obj.GetManagedFields() {
+//
+// This is a multi-step process.
+//
+// Step 1: All fields are owned by manager 'crossplane' operation 'Update'. This
+// represents all fields set by the claim or XR controller up to this point.
+//
+// Step 2: Upgrade is called for the first time. We delete all field managers.
+//
+// Step 3: The claim controller server-side applies its fully specified intent
+// as field manager 'apiextensions.crossplane.io/claim'. This becomes the
+// manager of all the fields that are part of the claim controller's fully
+// specified intent. All existing fields the claim controller didn't specify
+// become owned by a special manager - 'before-first-apply', operation 'Update'.
+//
+// Step 4: Upgrade is called for the second time. It deletes the
+// 'before-first-apply' field manager entry. Only the claim field manager
+// remains.
+//
+// Step 5: Eventually the XR reconciler updates a field (e.g. spec.resourceRefs)
+// and becomes owner of that field.
+func (u *PatchingManagedFieldsUpgrader) Upgrade(ctx context.Context, obj client.Object, ssaManager string, _ ...string) error {
+	// The XR doesn't exist, nothing to upgrade.
+	if !meta.WasCreated(obj) {
+		return nil
+	}
+
+	foundSSA := false
+	foundBFA := false
+	idxBFA := -1
+
+	for i, e := range obj.GetManagedFields() {
 		if e.Manager == ssaManager {
-			return nil
+			foundSSA = true
+		}
+		if e.Manager == "before-first-apply" {
+			foundBFA = true
+			idxBFA = i
 		}
 	}
-	p, err := csaupgrade.UpgradeManagedFieldsPatch(obj, sets.New[string](csaManagers...), ssaManager)
-	if err != nil {
-		return errors.Wrap(err, errCreatePatch)
-	}
-	if p == nil {
-		// No patch means there's nothing to upgrade.
+
+	switch {
+	// If our SSA field manager exists and the before-first-apply field manager
+	// doesn't, we've already done the upgrade. Don't do it again.
+	case foundSSA && !foundBFA:
 		return nil
+
+	// We found our SSA field manager but also before-first-apply. It should now
+	// be safe to delete before-first-apply.
+	case foundSSA && foundBFA:
+		p := []byte(fmt.Sprintf(`[{"op": "remove", "path": "/metadata/managedFields/%d"}]`, idxBFA))
```

Comment:
Does this patch needs to be made conditional on resourceVersion, so as to not overwrite/remove the incorrect  managed fields entry on the server, if they have changed before sending this patch ?

Reply:
I was under the impression that `client.Patch` the supplied `obj` to handle this automatically. Based on what `UpgradeManagedFieldsPatch` was doing, it seems like you're probably right though. Good catch!

https://github.com/kubernetes/client-go/blob/v0.30.0/util/csaupgrade/upgrade.go#L165

Reply:
Pushed a commit that addresses this.